packages/mongodb-explain-compat/lib/index.js

```diff
@@ -72,18 +72,67 @@ function mapStages(queryPlan, sbeExecutionStages) {
   });
 }
 
-module.exports = function(explain) {
-  explain = JSON.parse(JSON.stringify(explain));
-  if ((explain.explainVersion || 0) < 2) {
+function isAggregationExplain(explain) {
+  return !!explain.stages;
+}
+
+function isFirstStageCursorType(stages) {
+  return !!getStageCursorKey(stages[0]);
+}
+
+function getStageCursorKey(stage) {
+  return Object.keys(stage).find((x) => x.match(/cursor/i));
+}
+
+/**
+ * https://wiki.corp.mongodb.com/display/QUERY/Explain+Notes
+ */
+function mapAggregationExplain(explain) {
+  if (!explain.stages || explain.stages.length === 0) {
     return explain;
   }
-  delete explain.explainVersion;
+  // In aggregations for SBE, the cursor based stage is always first.
+  if (!isFirstStageCursorType(explain.stages)) {
+    return explain;
+  }
+
+  const firstStage = explain.stages[0];
+  const firstStageKey = getStageCursorKey(firstStage);
+  const winningPlan =
+    firstStage[firstStageKey].queryPlanner.winningPlan.queryPlan;
+  firstStage[firstStageKey].queryPlanner.winningPlan = winningPlan;
+
+  firstStage[firstStageKey].executionStats.executionStages = mapStages(
+    winningPlan,
+    firstStage[firstStageKey].executionStats.executionStages
+  );
+  firstStage[firstStageKey].queryPlanner.plannerVersion = 1;
+  return explain;
+}
+
+function mapQueryExplain(explain) {
   explain.queryPlanner.plannerVersion = 1;
   const winningPlan = explain.queryPlanner.winningPlan.queryPlan;
   explain.queryPlanner.winningPlan = winningPlan;
 
-  explain.executionStats.executionStages =
-    mapStages(winningPlan, explain.executionStats.executionStages);
+  explain.executionStats.executionStages = mapStages(
+    winningPlan,
+    explain.executionStats.executionStages
+  );
 
+  return explain;
+}
+
+module.exports = function (explain) {
+  explain = JSON.parse(JSON.stringify(explain));
+  if ((explain.explainVersion || 0) < 2) {
+    return explain;
+  }
+  delete explain.explainVersion;
+  if (isAggregationExplain(explain)) {
+    explain = mapAggregationExplain(explain);
+  } else {
+    explain = mapQueryExplain(explain);
+  }
```

Comment:
In https://github.com/mongodb/mongo/blob/e9e4e99d0ff87ec9b6ed2a689a4c42cdba7f5527/src/mongo/shell/explainable.js the outermost check is a sharding check, not an agg vs query check – which makes sense to me, because shard handling is always the last thing that happens.

Does this handle sharding properly? Can we add a test for that case as well?